File: backend/app/schemas/open_field.py
from __future__ import annotations
from typing import Optional, List, Any
from datetime import date, datetime
from pydantic import BaseModel


# ── PlantePere ────────────────────────────────────────────────────────────────

class PlantePereBase(BaseModel):
    id_variete: Optional[int] = None
    nom_libre:  Optional[str] = None
    notes:      Optional[str] = None

class PlantePereCreate(PlantePereBase):
    pass

class PlantePereUpdate(PlantePereBase):
    pass

class PlantePereRead(PlantePereBase):
    id_pere:     int
    id_projet:   int
    variete_nom: Optional[str] = None
    created_at:  Optional[datetime] = None

    class Config:
        from_attributes = True


# ── PlanteMere ────────────────────────────────────────────────────────────────

class PlanteMereBase(BaseModel):
    id_variete:             Optional[int]   = None
    nom_phenotype:          Optional[str]   = None
    id_plant:               Optional[int]   = None
    date_pollinisation:     Optional[date]  = None
    methode_pollinisation:  Optional[str]   = None   # naturelle | manuelle | pinceau
    pere_identifie:         Optional[bool]  = False
    id_pollen:              Optional[int]   = None
    nom_pere_libre:         Optional[str]   = None
    id_peres:               Optional[List[int]] = None   # IDs PlantePereOpenField probables
    notes_pollinisation:    Optional[str]   = None
    notes:                  Optional[str]   = None

class PlanteMereCreate(PlanteMereBase):
    pass

class PlanteMereUpdate(PlanteMereBase):
    pass

class RecolteInput(BaseModel):
    date_recolte:            Optional[date]  = None
    nb_graines:              Optional[int]   = None
    poids_graines_g:         Optional[float] = None
    qualite_graines:         Optional[str]   = None   # bonne | moyenne | immature
    # Pour créer automatiquement un PackGraine + nouvelle Variete
    creer_pack:              bool            = True
    nom_variete_croisement:  Optional[str]   = None   # nom de la nouvelle variété créée

class PlanteMereRead(PlanteMereBase):
    id_mere:         int
    id_projet:       int
    date_recolte:    Optional[date]  = None
    nb_graines:      Optional[int]   = None
    poids_graines_g: Optional[float] = None
    qualite_graines: Optional[str]   = None
    id_packgraine:   Optional[int]   = None
    created_at:      Optional[datetime] = None
    # Enrichissements
    variete_nom:     Optional[str]   = None
    pollen_nom:      Optional[str]   = None
    plant_label:     Optional[str]   = None
    peres_labels:    Optional[List[str]] = None  # noms des pères probables

    class Config:
        from_attributes = True


# ── ProjetOpenField ───────────────────────────────────────────────────────────

class ProjetOpenFieldBase(BaseModel):
    nom:          str
    saison:       Optional[str]  = None
    lieu:         Optional[str]  = None
    conditions:   Optional[str]  = None   # outdoor | greenhouse | guerrilla
    id_culture:   Optional[int]  = None
    statut:       Optional[str]  = "planifie"
    description:  Optional[str]  = None
    notes:        Optional[str]  = None

class ProjetOpenFieldCreate(ProjetOpenFieldBase):
    pass

class ProjetOpenFieldUpdate(BaseModel):
    nom:          Optional[str]  = None
    saison:       Optional[str]  = None
    lieu:         Optional[str]  = None
    conditions:   Optional[str]  = None
    id_culture:   Optional[int]  = None
    statut:       Optional[str]  = None
    description:  Optional[str]  = None
    notes:        Optional[str]  = None

class ProjetOpenFieldRead(ProjetOpenFieldBase):
    id_projet:    int
    created_at:   Optional[datetime] = None
    meres:        List[PlanteMereRead] = []
    peres:        List[PlantePereRead] = []
    # Enrichissements
    culture_nom:  Optional[str]  = None
    nb_meres:     int            = 0
    nb_peres:     int            = 0
    nb_graines_total: int        = 0

    class Config:
        from_attributes = True

File: frontend/src/api/openField.ts
import client from './client'

// ── Types ─────────────────────────────────────────────────────────────────────

export type StatutProjet = 'planifie' | 'en_cours' | 'recolte' | 'termine'
export type ConditionsProjet = 'outdoor' | 'greenhouse' | 'guerrilla'
export type MethodePollinisation = 'naturelle' | 'manuelle' | 'pinceau'
export type QualiteGraines = 'bonne' | 'moyenne' | 'immature'

export interface PlanteMere {
  id_mere: number
  id_projet: number
  id_variete?: number
  variete_nom?: string
  nom_phenotype?: string
  id_plant?: number
  plant_label?: string
  date_pollinisation?: string
  methode_pollinisation?: MethodePollinisation
  pere_identifie: boolean
  id_pollen?: number
  pollen_nom?: string
  nom_pere_libre?: string
  notes_pollinisation?: string
  date_recolte?: string
  nb_graines?: number
  poids_graines_g?: number
  qualite_graines?: QualiteGraines
  id_peres?: number[]
  id_packgraine?: number
  peres_labels?: string[]
  notes?: string
  created_at?: string
}

export interface ProjetOpenField {
  id_projet: number
  nom: string
  saison?: string
  lieu?: string
  conditions?: ConditionsProjet
  id_culture?: number
  culture_nom?: string
  statut: StatutProjet
  description?: string
  notes?: string
  created_at?: string
  meres: PlanteMere[]
  peres: PlantePere[]
  nb_meres: number
  nb_peres: number
  nb_graines_total: number
}

export interface ProjetCreate {
  nom: string
  saison?: string
  lieu?: string
  conditions?: ConditionsProjet
  id_culture?: number
  statut?: StatutProjet
  description?: string
  notes?: string
}

export interface PlanteMereCreate {
  id_variete?: number
  nom_phenotype?: string
  id_plant?: number
  date_pollinisation?: string
  methode_pollinisation?: MethodePollinisation
  pere_identifie?: boolean
  id_pollen?: number
  nom_pere_libre?: string
  id_peres?: number[]
  notes_pollinisation?: string
  notes?: string
}

export interface RecolteInput {
  date_recolte?: string
  nb_graines?: number
  poids_graines_g?: number
  qualite_graines?: QualiteGraines
  creer_pack?: boolean
  nom_variete_croisement?: string
}


export interface PlantePere {
  id_pere:     number
  id_projet:   number
  id_variete?: number
  variete_nom?: string
  nom_libre?:  string
  notes?:      string
  created_at?: string
}

export interface PlantePereCreate {
  id_variete?: number
  nom_libre?:  string
  notes?:      string
}

// ── API ───────────────────────────────────────────────────────────────────────

export const openFieldAPI = {
  // Projets
  list:   (statut?: string) =>
    client.get<ProjetOpenField[]>('/open-field', { params: statut ? { statut } : {} }),
  get:    (id: number) =>
    client.get<ProjetOpenField>(`/open-field/${id}`),
  create: (data: ProjetCreate) =>
    client.post<ProjetOpenField>('/open-field', data),
  update: (id: number, data: Partial<ProjetCreate>) =>
    client.patch<ProjetOpenField>(`/open-field/${id}`, data),
  delete: (id: number) =>
    client.delete(`/open-field/${id}`),

  // Mères
  addMere:    (projetId: number, data: PlanteMereCreate) =>
    client.post<ProjetOpenField>(`/open-field/${projetId}/meres`, data),
  updateMere: (projetId: number, mereId: number, data: Partial<PlanteMereCreate>) =>
    client.patch<ProjetOpenField>(`/open-field/${projetId}/meres/${mereId}`, data),
  deleteMere: (projetId: number, mereId: number) =>
    client.delete(`/open-field/${projetId}/meres/${mereId}`),

  // Pères
  addPere:    (projetId: number, data: PlantePereCreate) =>
    client.post<ProjetOpenField>(`/open-field/${projetId}/peres`, data),
  updatePere: (projetId: number, pereId: number, data: Partial<PlantePereCreate>) =>
    client.patch<ProjetOpenField>(`/open-field/${projetId}/peres/${pereId}`, data),
  deletePere: (projetId: number, pereId: number) =>
    client.delete(`/open-field/${projetId}/peres/${pereId}`),

  // Récolte
  recolte: (projetId: number, mereId: number, data: RecolteInput) =>
    client.post<ProjetOpenField>(`/open-field/${projetId}/meres/${mereId}/recolte`, data),
}
